File: Merge.fs
open System

// takes the first n from a list
let take n list = 
    List.ofSeq <| Seq.take n list

// takes everything after n of a list
let rec takeRest n list = 
    match (n, list) with
    // if n is zero take all
    | (0, l) -> l
    // recursive part
    | (n, _ :: tl) -> takeRest (n - 1) tl
    | (_, _) -> failwith "unknown pattern"

let rec mergeList list1 list2 =
    match (list1, list2) with
    // if both lists are empty return empty
    | [], [] -> []
    // if only one list is empty return the other one
    | [], l -> l
    | l, [] -> l
    // tail end recursion
    | hd1 :: tl1, (hd2 :: _ as tl2) when hd1 < hd2 -> hd1 ::  mergeList tl1 tl2
    | l1, hd2 :: tl2 -> hd2 :: mergeList l1 tl2

let rec mergeSort = function
    // if empty or only one element
    | [] -> [] 
    | [x] -> [x]
    // else
    | l -> 
        // get midpoint
        let n = (int)(List.length l / 2)      
        // mergesort left and right side  
        let list1 = mergeSort (take n l)
        let list2 = mergeSort (takeRest n l)
        // then merge
        mergeList list1 list2


let genRandomNumbers count =
    let rnd = System.Random()
    List.init count (fun _ -> rnd.Next (1000001))

[<EntryPoint>]
let main args = 
    for i = 1 to 50 do
        let input = genRandomNumbers 10000
        let watch = new System.Diagnostics.Stopwatch()
        watch.Start()
        let result = mergeSort input
        watch.Stop()
        Console.WriteLine("{0}" , watch.ElapsedMilliseconds)
    0

File: Merge.cs
using System;
					
public class Program
{
	private static void Merge(int[] input, int left, int middle, int right)
	{
		// creates new array of ints for the left and right portion
		int[] leftArray = new int[middle - left + 1];
		int[] rightArray = new int[right - middle];
	
		// copy all the values into the respective array
		Array.Copy(input, left, leftArray, 0, middle - left + 1);
		Array.Copy(input, middle + 1, rightArray, 0, right - middle);
	
		int i = 0;
		int j = 0;
		// actual merging
		for (int k = left; k < right + 1; k++)
		{
			if (i == leftArray.Length)
			{
				input[k] = rightArray[j];
				j++;
			}
			else if (j == rightArray.Length)
			{
				input[k] = leftArray[i];
				i++;
			}
			else if (leftArray[i] <= rightArray[j])
			{
				input[k] = leftArray[i];
				i++;
			}
			else
			{
				input[k] = rightArray[j];
				j++;
			}
		}
	}
 
	private static void MergeSort(int[] input, int left, int right)
	{
		if (left < right)
		{
			int middle = (left + right) / 2;
	
			// merge sort both sides
			MergeSort(input, left, middle);
			MergeSort(input, middle + 1, right);
	
			// rearrange
			Merge(input, left, middle, right);
		}
	}    
	
	public static void Main()
	{
		int [] testValues = new int [10] {100,10,56,896,145,4512,16456,1,568,115615};

		MergeSort(testValues, 0, testValues.Length - 1);
	}
}